corehq/apps/reports/standard/sms.py

```diff
@@ -301,11 +326,430 @@ def rows(self):
+            display_text,
+        )
+        return self.table_cell(display_text, display)
+
+    def get_survey_detail_link(self, subevent):
+        form_name = subevent.form_name or _('Unknown')
+        if not subevent.xforms_session_id:
+            return self._fmt(form_name)
+        else:
+            display = '<a target="_blank" href="/a/%s/reports/survey_detail/?id=%s">%s</a>' % (
+                self.domain,
+                subevent.xforms_session_id,
+                form_name,
+            )
+            return self.table_cell(form_name, display)
+
+
+class MessagingEventsReport(BaseMessagingEventReport):
+    name = ugettext_noop('Past Events')
+    slug = 'messaging_events'
+    fields = [
+        DatespanFilter,
+        EventTypeFilter,
+    ]
+
+    @property
+    def headers(self):
+        header = DataTablesHeader(
+            DataTablesColumn(_('Date')),
+            DataTablesColumn(_('Content')),
+            DataTablesColumn(_('Type')),
+            DataTablesColumn(_('Recipient')),
+            DataTablesColumn(_('Status')),
+            DataTablesColumn(_('Detail')),
+        )
+        header.custom_sort = [[0, 'desc']]
+        return header
+
+    def get_filters(self):
+        source_filter = []
+        content_type_filter = []
+        event_type_filter = EventTypeFilter.get_value(self.request, self.domain)
+
+        for source_type, x in MessagingEvent.SOURCE_CHOICES:
+            if source_type in event_type_filter:
+                if source_type == MessagingEvent.SOURCE_OTHER:
+                    source_filter.extend([
+                        MessagingEvent.SOURCE_OTHER,
+                        MessagingEvent.SOURCE_FORWARDED,
+                    ])
+                else:
+                    source_filter.append(source_type)
+
+        for content_type, x in MessagingEvent.CONTENT_CHOICES:
+            if content_type in event_type_filter:
+                if content_type == MessagingEvent.CONTENT_SMS_SURVEY:
+                    content_type_filter.extend([
+                        MessagingEvent.CONTENT_SMS_SURVEY,
+                        MessagingEvent.CONTENT_IVR_SURVEY,
+                    ])
+                else:
+                    content_type_filter.append(content_type)
+
+        return (source_filter, content_type_filter)
+
+    @property
+    def rows(self):
+        source_filter, content_type_filter = self.get_filters()
+
+        # We need to call distinct() on this because it's doing an
+        # outer join to sms_messagingsubevent in order to filter on
+        # subevent content types.
+        data = MessagingEvent.objects.filter(
+            Q(domain=self.domain),
+            Q(date__gte=self.datespan.startdate_utc),
+            Q(date__lte=self.datespan.enddate_utc),
+            (Q(source__in=source_filter) |
+                Q(content_type__in=content_type_filter) |
+                Q(messagingsubevent__content_type__in=content_type_filter)),
+        ).distinct()
```

Comment:
Ideally, yes, but I'd prefer to add that in a separate PR as this one is already quite large.